roms_tools/setup/cdr_forcing.py

```diff
@@ -21,171 +18,210 @@
+                "Each release must have a unique name."
+            )
 
-        for tracer, val in self.tracer_concentrations.items():
+        return self
 
-            if isinstance(val, list):
-                if len(val) != num_times:
-                    raise ValueError(
-                        f"The length of tracer '{tracer}' ({len(val)}) does not match the number of times ({num_times})."
-                    )
+    def determine_release_type(self) -> "ReleaseCollector":
+        """Ensure all releases are of the same type, and set the release_type."""
+        if not all(
+            isinstance(release, type(self.releases[0])) for release in self.releases
+        ):
+            raise ValueError("Not all releases are of the same type.")
 
-        return self
+        return type(self.releases[0])
 
-    @model_validator(mode="after")
-    def extend_to_endpoints(self) -> "Release":
-        """Ensure that the release time series starts at `start_time` and ends at
-        `end_time`.
 
-        If `volume_fluxes` is a list and does not cover the endpoints, zero volume fluxes are added.
-        Tracer concentrations are extended accordingly by duplicating endpoint values.
+class CDRForcingDatasetBuilder:
+    """Constructs the xarray `Dataset` to be saved as NetCDF."""
 
-        If `times` is empty, [start_time, end_time] is used with zero fluxes and duplicated tracer values.
-        """
-        if not self.times:
-            self.times = [self.start_time, self.end_time]
-            self.volume_fluxes = [self.volume_fluxes, self.volume_fluxes]
-            self.tracer_concentrations = {
-                tracer: [vals, vals]
-                if not isinstance(vals, list)
-                else [vals[0], vals[0]]
-                for tracer, vals in self.tracer_concentrations.items()
+    def __init__(self, releases, model_reference_date, release_type):
+        self.releases = releases
+        self.model_reference_date = model_reference_date
+        self.release_type = release_type
+
+    def build(self) -> xr.Dataset:
+        all_times = [np.array(r.times, dtype="datetime64[ns]") for r in self.releases]
+        union_times = np.unique(np.concatenate(all_times))
+        union_rel_times = [
+            convert_to_relative_days(t, self.model_reference_date) for t in union_times
+        ]
+
+        ds = xr.Dataset()
+        ds["time"] = ("time", union_times)
+        ds["cdr_time"] = ("time", union_rel_times)
+        ds["cdr_lon"] = ("ncdr", [r.lon for r in self.releases])
+        ds["cdr_lat"] = ("ncdr", [r.lat for r in self.releases])
+        ds["cdr_dep"] = ("ncdr", [r.depth for r in self.releases])
+        ds["cdr_hsc"] = ("ncdr", [r.hsc for r in self.releases])
+        ds["cdr_vsc"] = ("ncdr", [r.vsc for r in self.releases])
+        ds = ds.assign_coords(
+            {"release_name": (["ncdr"], [r.name for r in self.releases])}
+        )
+
+        # Assign attributes
+        attr_map = {
```

Comment:
consider pulling this out into a class method to shorten the monster!

```
@classmethod
def get_attr_map(cls, reference_date: int) -> t.Dict[str, t.Dict[str, str]]:
    return { ... attr_map here ... }
```

Reply:
Sounds good. Done in a5f12ae (except that it's  a method, not a classmethod).